feat(orderBy): add support for custom comparators

gkalpak opened the pull request:
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
Feature

**What is the current behavior? (You can also link to an open issue here)**
The comparator used by `orderBy` is hard-coded. 

**What is the new behavior (if this is a feature change)?**
The user can specify a custom comparator function.

**Does this PR introduce a breaking change?**
No

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**Other information**:
With this change the 3rd argument is interpreted as a comparator function, used to compare the values returned by the predicates. Leaving it empty falls back to the default comparator and a value of `false` is a special case that also falls back to the default comparator but reverses the order. Thus the new implementation is backwards compatible.

This commit also expands the documentation to cover the algorithm used to sort elements and adds a few more unit and e2e tests (unrelated to the change).

Helps with #12572 (maybe this is as close as we want to get).

Fixes #13238
Fixes #14455

Closes #5123
Closes #8112
Closes #10368

gkalpak committed:
feat(orderBy): add support for custom comparators

With this change the 3rd argument is interpreted as a comparator function, used to compare the
values returned by the predicates. Leaving it empty falls back to the default comparator and a
value of `false` is a special case that also falls back to the default comparator but reverses the
order. Thus the new implementation is backwards compatible.

This commit also expands the documentation to cover the algorithm used to sort elements and adds
a few more unit and e2e tests (unrelated to the change).

Helps with #12572 (maybe this is as close as we want to get).

Fixes #13238
Fixes #14455

Closes #5123
Closes #8112
Closes #10368

Narretz commented on test/ng/filter/orderBySpec.js:
```diff
@@ -157,6 +183,38 @@ describe('Filter: orderBy', function() {
     });
 
 
+    it('should support a custom comparator', function() {
```
I feel that the test is a bit too clever with "ownerD" being the nerd in question. ;)

gkalpak replied:
I would go as far as calling it "nerdy" :stuck_out_tongue: 

I added a bunch of other tests, so it's not so necessary any more - I just kept it around because I liked it. I'm fine removing it if you think it's too confusing.

Narretz commented:
What if you want to have a custom comparator and the option to easily reverse the order?

gkalpak commented:
This now ready for review :smile:

I updated the docs and added more tests. Especially wrt to _how_ we sort things (beyond the: _"numbers numerically, strings alphabetically"_). I believe that if people are to rely on some ordering algorithm, they have the right to know what that is.

Narretz commented on src/ng/filter/orderBy.js:
```diff
@@ -6,44 +6,118 @@
+ * order (expression is set to `'-age'`). The `comparator` is not set, which means it defaults to
+ * the built-in comparator.
+ *
+   <example module="orderByExample1">
```
Can you name this and the other examples (name="desriptiveName"). That makes them easier to find in the build folder.

Narretz commented on src/ng/filter/orderBy.js:
```diff
@@ -6,44 +6,118 @@
+ * The `collection` can be an Array or array-like object (e.g. NodeList, jQuery object, TypedArray,
+ * String, etc).
+ *
+ * The `expression` can be a signle predicate or a list of predicates, each serving as a tie-breaker
```
signle => single
and I would add a comma after or, because at the moment, it's not super clear to what "each" refers

Narretz commented on src/ng/filter/orderBy.js:
```diff
@@ -6,44 +6,118 @@
+ * In order to ensure that the sorting will be deterministic across platforms, if none of the
+ * specified predicates can distinguish between two items, `orderBy` will automatically introduce a
+ * dummy predicate that returns the item's index as `value`.
+ * (If you are using a custom comparator, make sure it can handle this case as well.)
```
"as well" is redundant, imo

gkalpak replied:
I changed it to: _"make sure it can handle this predicate as well"_

Narretz commented on src/ng/filter/orderBy.js:
```diff
@@ -6,44 +6,118 @@
+ *
+ * 1. If the object has a `valueOf()` method that returns a primitive, its return value will be
+ *    used instead.<br />
+ *    (If the object does have a `valueOf()` method, but it returns another object, then the
```
", but it returns" -> "that returns"

gkalpak replied:
Changed it to: _"If the object has a `valueOf()` method that returns..."_

Narretz commented on src/ng/filter/orderBy.js:
```diff
@@ -6,44 +6,118 @@
+ *    returned object will be used in subsequent steps.)
+ * 2. If the object has a custom `toString()` method (i.e. not the one inherited from `Object`) that
+ *    returns a primitive, its return value will be used instead.<br />
+ *    (If the object does have a `toString()` method, but it returns another object, then the
```
same as above

gkalpak replied:
Same as above.

Narretz commented on src/ng/filter/orderBy.js:
```diff
@@ -6,44 +6,118 @@
+ * for comparing with other items.
+ *
+ * The comparison is done using the `comparator` function. If none is specified, a default, built-in
+ * comparator is used (see below for details). Passing `true` as the comparator has a special
```
I think it makes sense to say here briefly that strings are sorted alphabetically and numbers numerically (the most common use cases)

Narretz commented on src/ng/filter/orderBy.js:
```diff
@@ -6,44 +6,118 @@
+ * ### The default comparator
+ *
+ * The default, built-in comparator should be sufficient for most usecases. In short, it compares
+ * numbers numerically, strings alphabetically (and case-insensitively), for objects falls back to
```
Does that mean if the predicate is an object, the default ordering maintains the original order?

gkalpak replied:
If the predicate returns an object for two elements, there these two elements will retain their original relative ordering.
So, yes :smiley:

Narretz commented on src/ng/filter/orderBy.js:
```diff
@@ -6,44 +6,118 @@
- *      element itself is used to compare where sorting.
- *    - `Array`: An array of function or string predicates. The first predicate in the array
- *      is used for sorting, but when two items are equivalent, the next predicate is used.
+ *    - `Function`: A getter function. Ths function will be called with each item as argument and
```
Ths => this

Narretz commented on src/ng/filter/orderBy.js:
```diff
@@ -6,44 +6,118 @@
+ *    - `string`: An Angular expression. This expression will be evaluated against each item and the
+ *      result will be used for sorting. For example, use `name` to sort by a property called `name`
+ *      or `name.substring(0, 3)` to sort by the first 3 characters of the `name` property.<br />
+ *      (The result of a constant expression is interpreted as a property name to be used for
```
Isn't that the same as in the sentence above? ( For example, use `name` to sort by a property called `name`)

gkalpak replied:
There is a subtle difference:

If the Angular expression `name` is evaluated against an object, it will be evaluated to the object's `name` proeprty. (This is standard Angular behavior.)
If the Angular expression `"special name"` is evaluated against an object, typically it would be evaluated to the string `"special name"` (because it's a constant expression - it doesn't access the evluation context). But here, it is special cased: It will actually evaluate to the object's `special name` property.

I guess it was introduced back when `this` was not an lias to the evaluation context in Angular expressions. Nowadays, you can use `this['special name']`, so we don't need the special treating, but I guess we need the backwards compatibility, so...

Narretz replied:
I feel like this doesn't really explain that "name" is a property on the scope, and that the value of that property determines which property of the item is used to order the collection.

gkalpak replied:
That's not what it is supposed to explain :smiley: It is supposed to explain what the string `name` (i.e. `'name'`) would work.

I admit this is not super clear - hopefully the examples clear things up. I added the surrounding quotes and renamed `name` to `label`, to avoid the confusion of _"the `name` property"_ vs _"the property name"_ (i.e. the name of some property).
We can always improve the docs once merged :smiley:

Narretz commented on src/ng/filter/orderBy.js:
```diff
@@ -103,100 +210,332 @@
+ *
+ * @example
+ * If you have very specific requirements about the way items are sorted, you can pass your own
+ * comparator function. For example, you might need to compare some strings is a locale-sensitive
```
is => in

Narretz commented:
Can you also mention in the docs that orderBy can only order alphabetically by latin characters? Actually, which characters are supported anyway?

gkalpak committed:
fixup 1

gkalpak commented:
@Narretz, I have addressed your comments. PTAL

Regarding the "supported" characters:
It uses the default `<` comparison (so it's not latin characters only). It is just that it doesn't do locale-sensitive comparison (and in some locales it makes a difference). The "locale-insensitive-ness" of the default comparator is already mentioned - other than that all characters are supported (in a sense).

Narretz commented:
Thanks for the update.
I remember we talked about how to keep the API backwards compatible in a sane way. Did you change anything about this since?

gkalpak commented:
This is already backwards compatible - but you are right we did talk about something:
Making it easy to have a custom comparator and still be able to specify ascending/descending order.

We settled for `orderBy(collection[, expression[, reverse[, comparator]]])`. I'll make the changes.

gkalpak committed:
fixup 2

- Decouple `reverse` from `comparator` arguments.

gkalpak commented:
@Narretz, I moved `comparator` as a 4th argument. Now you can specify `reverse` and `comparator` independently. PTAL

Narretz commented on src/ng/filter/orderBy.js:
```diff
@@ -6,44 +6,121 @@
- * @returns {Array} Sorted copy of the source array.
+ * @param {boolean=} reverse - If `true`, reverse the sorting order.
+ * @param {(Function)=} comparator - The comparator function used to determine the relative order of
+ *    value pairs. If omitted, a built-in comparator will be used.
```
Is there more than one built-in comparator? Or should it be "the default comparator"?

gkalpak replied:
There's only one. Changed it to `the`.

Narretz commented on src/ng/filter/orderBy.js:
```diff
@@ -103,100 +212,333 @@
+ * If you have very specific requirements about the way items are sorted, you can pass your own
+ * comparator function. For example, you might need to compare some strings in a locale-sensitive
+ * way. (When specifying a custom comparator, you also need to pass a value for the `reverse`
+ * argument - passing `false` retains the default sorting order, i.e. ascending.)
```
Actually, all falsy values. Specifically, undefined can be used (and I would use that if I don't care about a parameter).

gkalpak replied:
We have documented this as an (optional) boolean. So, although we treat all falsy values as `false` internally, I prefer to use a boolean in the example.

Narretz commented on src/ng/filter/orderBy.js:
```diff
@@ -6,44 +6,121 @@
- * as expected, make sure they are actually being saved as numbers and not strings.
- * Array-like values (e.g. NodeLists, jQuery objects, TypedArrays, Strings, etc) are also supported.
+ * Returns an array containing the items from the specified `collection`, ordered by a `comparator`
+ * function based on the values computed using the `expression` predicate.
```
A small example of what the result of the expression might be would be good (in most cases it's the property of an object)

Narretz commented:
I left a few smaller comments where I think the docs could be clarified. I've also opened a PR in your repo that adds descriptions ot the examples
Otherwise LGTM. Maybe @petebacondarwin can give it another look, too.

petebacondarwin commented on test/ng/filter/orderBySpec.js:
```diff
@@ -237,6 +218,346 @@ describe('Filter: orderBy', function() {
+
+
+    describe('(reversing order)', function() {
+      it('should not reverse collection if `comparator` param is falsy',
```
`comparator` param -> `reverse` param

petebacondarwin commented on test/ng/filter/orderBySpec.js:
```diff
@@ -237,6 +218,346 @@ describe('Filter: orderBy', function() {
+      );
+
+
+      it('should reverse collection if `comparator` param is `true`, even without an `expression`',
```
`comparator` param -> `reverse` param

petebacondarwin commented:
Apart from the couple of nitpicks in docs and tests, this LGTM.
Don't forget to fix up the commit message.

gkalpak committed:
fixup 3

- Fix typos and wording.

gkalpak commented:
Fixed and merged! Thx @Narretz and @petebacondarwin for the reviews :+1:
